lib/jsonWebToken.js

```diff
@@ -0,0 +1,195 @@
+ */
+
+'use restrict';
+
+const jws = require('jws');
+
+/* Verify the token and return the payload
+ *
+ * @jwtString
+ * @PEMKey
+ * @options
+ *   - algorithms (required)
+ *   - audience (required)
+ *   - issuer (optional, validate if provided)
+ *   - subject (optional, validate if provided)
+ *   - ignoreExpiration (optional, if not set true we will validate expiration)
+ * @callback
+ */
+exports.verify = function(jwtString, PEMKey, options, callback) {
+
+  /*********************************************************************
+   * Checking parameters
+   ********************************************************************/
+
+  // check the existence of callback function and options, if we don't have them, that means we have
+  // less than 4 parameters passed. This is a server (code) error, we should throw.
+  if (!callback)
+    throw new Error('callback must be provided in jsonWebToken.verify');
+  if (typeof callback !== 'function')
+    throw new Error('callback in jsonWebToken.verify must be a function');
+  if (!options)
+    throw new Error('options must be provided in jsonWebToken.verify');
+
+  // check jwtString and PEMKey are provided. Since jwtString and PEMKey are generated, this is
+  // a non-server error, we shouldn't throw, we just give the error back and let authentication fail.
+  if (!jwtString || jwtString === '')
+    return done(new Error('jwtString must be provided in jsonWebToken.verify'));
+  if (!PEMKey || PEMKey === '')
+    return done(new Error('PEMKey must be provided in jsonWebToken.verify'));
+
+  // asynchronous wrapper for callback
+  var done = function() {
+    var args = Array.prototype.slice.call(arguments, 0);
+    return process.nextTick(function() {
+      callback.apply(null, args);
+    });
+  };
+
+  // make sure we have the required fields in options
+  if (typeof options.audience !== 'string' || options.audience === '')
+    return done(new Error('options.audience is missing in jsonWebToken.verify'));
+  if (!options.algorithms)
+    return done(new Error('options.algorithms is missing in jsonWebToken.verify'));
+  if (!Array.isArray(options.algorithms) || options.algorithms.length == 0 ||
+    (options.algorithms.length == 1 && options.algorithms[0] === 'none'))
+    return done(new Error('options.algorithms must be an array containing at least one algorithm'));
+
+  /*********************************************************************
+   * Checking jwtString structure, getting header, payload and signature
+   ********************************************************************/
+
+  // split jwtString, make sure we have three parts and we have signature
+  var parts = jwtString.split('.');
+  if (parts.length !== 3)
+    return done(new Error('jwtString is malformed'));
+  if (parts[2] === '')
+    return done(new Error('signature is missing in jwtString'));
+  
+  // decode jwsString
+  var decodedToken;
```

Comment:
@tushargupta51 later at line 125 we need to use the original string to validate signature.